frontend/src/components/node/node.jsx

```diff
@@ -34,85 +33,113 @@ const CursorNode = ({ nodeId }) => {
-        if (direction === NODE.OUTPUT) {
-            setValidHoverClassLabel(state?.selectedOutput === null ? "valid-hover" : "");
-        }
-        if (direction === NODE.INPUT) {
-            setValidHoverClassLabel(state?.selectedOutput !== null ? "valid-hover" : "")
+const InputNode = ({ nodeId, selectedOutput, dispatch, className, setInputs }) => {
+    const handleMouseUp = (e) => {
+        if (selectedOutput !== null) {
+            dispatch({
+                type: 'ADD_CONNECTION',
+                payload: { end: `${nodeId}`, setInputs },
+            });
         }
     };
 
+    return (
+        <div
+            className={className}
+            onMouseUp={handleMouseUp}
+            id={`${nodeId}`}
+        >{'>'}</div>
+    );
+};
 
-
+const OutputNode = ({ nodeId, dispatch, className }) => {
     const handleMouseDown = (e) => {
-        e.preventDefault();  // catch the mouse down here
-
+        e.preventDefault(); // catch the mouse down here
+        const CURSOR_SHIFT_HACK = 100;
         const rect = e.target.getBoundingClientRect();
         /* get the x, y offset from the center of the node */
-        const offsetX = e.clientX - rect.left - ((rect.right - rect.left) / 2);
-        const offsetY = e.clientY - rect.top - ((rect.bottom - rect.top) / 2);
-
-        if (direction === NODE.OUTPUT) {
-            dispatch({
-                type: 'START_CURSOR_NODE_DRAG',
-                payload: {
-                    start: `node-${nodeId}`,
-                    end: `cursor-node-${nodeId}`,
-                    offsetX,
-                    offsetY
-                }
-            })
-        }
+        const offsetX = (e.clientX - rect.left - (rect.right - rect.left) / 2) + CURSOR_SHIFT_HACK;
+        const offsetY = (e.clientY - rect.top - (rect.bottom - rect.top) / 2) + CURSOR_SHIFT_HACK;
+
+        dispatch({
+            type: 'START_CURSOR_NODE_DRAG',
+            payload: {
+                start: `${nodeId}`,
+                end: `cursor-${nodeId}`,
+                offsetX,
+                offsetY,
+            },
+        });
     };
 
 
-    const handleMouseUp = (e) => {
-        if ((direction === NODE.INPUT) && (state?.selectedOutput !== null)) {
-            dispatch({
-                type: 'ADD_CONNECTION',
-                payload: { end: `node-${nodeId}` }
-            })
-        }
+    return (
+        <>
+            <div className={className} id={`${nodeId}`}>{'>'}</div>
+            <CursorNode nodeId={nodeId} handleMouseDown={handleMouseDown} />
+        </>
+    );
+};
+
+const Node = ({ direction, setNodeId, setInputs }) => {
+    const [validHoverClassLabel, setValidHoverClassLabel] = useState('');
+    const nodeIdRef = useRef(useNextNodeId());    
+    useEffect(() => {
+        setNodeId(nodeIdRef.current);
+    }, []);
+
+    const globalState = useContext(store);
+    const { dispatch, state } = globalState;
+    const { selectedOutput } = state;

```

Comment:
Probably remove this and keep it namespaced under state so it's clear where it's coming from.